src/ccache.cpp

```diff
@@ -1590,7 +1608,14 @@ calculate_result_name(Context& ctx,
     //     share manifests and a/r.h exists.
     // * The expansion of __FILE__ may be incorrect.
     hash.hash_delimiter("inputfile");
-    hash.hash(ctx.args_info.input_file);
+
+    // If base_dir is set, use the relative path to generate hash
+    if (!ctx.config.base_dir().empty() && Util::is_absolute_path(ctx.args_info.input_file)) { 
+      std::string hash_path = Util::get_relative_path(ctx.config.base_dir(), ctx.args_info.input_file);
+      hash.hash(hash_path);
+    } else {
+        hash.hash(ctx.args_info.input_file);
+    }
```

Comment:
This will make the expansion of the `__FILE__` macro incorrect (see line 1609). There used to be a way of opting out of hashing the input file (see 02bcef0d82ac8f1397c9b13bf1215cf770e6b7da for background) so it's a deliberate choice to include the full input path in the hash.

That said, I think I would be OK with making the input path relative if the "`file_macro` sloppiness" setting is reintroduced but with new semantics:

```
if file_macro sloppiness is set:
    input_file = make_relative(input_file)
hash(input_file)
```
That is, keep `input_file` as is by default but make it relative if `file_macro` sloppiness is set.